src/rust/engine/nailgun/src/server.rs

```diff
@@ -0,0 +1,355 @@
+
+#![deny(warnings)]
+// Enable all clippy lints except for many of the pedantic ones. It's a shame this needs to be copied and pasted across crates, but there doesn't appear to be a way to include inner attributes from a common source.
+#![deny(
+  clippy::all,
+  clippy::default_trait_access,
+  clippy::expl_impl_clone_on_copy,
+  clippy::if_not_else,
+  clippy::needless_continue,
+  clippy::unseparated_literal_suffix,
+// TODO: Falsely triggers for async/await:
+//   see https://github.com/rust-lang/rust-clippy/issues/5360
+// clippy::used_underscore_binding
+)]
+// It is often more clear to show that nothing is being moved.
+#![allow(clippy::match_ref_pats)]
+// Subjective style.
+#![allow(
+  clippy::len_without_is_empty,
+  clippy::redundant_field_names,
+  clippy::too_many_arguments
+)]
+// Default isn't as big a deal as people seem to think it is.
+#![allow(clippy::new_without_default, clippy::new_ret_no_self)]
+// Arc<Mutex> can be more clear than needing to grok Orderings:
+#![allow(clippy::mutex_atomic)]
+
+use nails::execution::{self, sink_for, stream_for, ChildInput, ChildOutput, ExitCode};
+use nails::{Child, Nail};
+use task_executor::Executor;
+use tokio::fs::File;
+use tokio::net::TcpListener;
+use tokio::sync::{Notify, RwLock};
+
+use std::collections::HashMap;
+use std::io;
+use std::net::Ipv4Addr;
+use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd, RawFd};
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use bytes::Bytes;
+use futures::channel::{mpsc, oneshot};
+use futures::{future, sink, stream, FutureExt, SinkExt, StreamExt, TryStreamExt};
+use log::{debug, error, info};
+
+pub struct Server {
+  exit_sender: oneshot::Sender<()>,
+  exited_receiver: oneshot::Receiver<Result<(), String>>,
+  port: u16,
+}
+
+impl Server {
+  ///
+  /// Spawn the server on a background Task.
+  ///
+  /// The port provided here may be `0` in order to request a random port. A caller can use
+  /// `Server.port()` to determine what port was actually selected.
+  ///
+  pub async fn new(
+    executor: Executor,
+    port_requested: u16,
+    runner: impl RawFdRunner + 'static,
+  ) -> Result<Server, String> {
+    let listener = TcpListener::bind((Ipv4Addr::new(127, 0, 0, 1), port_requested))
+      .await
+      .map_err(|e| format!("Could not bind to port {}: {:?}", port_requested, e))?;
+    let port_actual = listener
+      .local_addr()
+      .map_err(|e| format!("No local address for listener: {:?}", e))?
+      .port();
+
+    // NB: The C client requires noisy_stdin (see the `nails` crate for more info), but neither
+    // `nails` nor the pants python client do.
+    let config = nails::Config::default().noisy_stdin(false);
+    let nail = RawFdNail {
+      executor: executor.clone(),
+      runner: Arc::new(runner),
+    };
+
+    let (exited_sender, exited_receiver) = oneshot::channel();
+    let (exit_sender, exit_receiver) = oneshot::channel();
+
+    let _join = executor.spawn(Self::serve(
```

Comment:
When does this future get `.await`ed? Do we expect it to complete on its own?

Reply:
This is code @stuhood wrote that I'm just moving around, but I think the idea is that `serve` should execute continuously until the `pantsd` process itself is shut down.